src/arch/riscv/interrupts.hh

```diff
@@ -137,83 +82,32 @@ class Interrupts : public BaseInterrupts
         return checkNonMaskableInterrupt() || (ip & ie & globalMask()).any();
     }
 
-    Fault
-    getInterrupt() override
-    {
-        assert(checkInterrupts());
-        if (checkNonMaskableInterrupt())
-            return std::make_shared<NonMaskableInterruptFault>();
-        std::bitset<NumInterruptTypes> mask = globalMask();
-        const std::vector<int> interrupt_order {
-            INT_EXT_MACHINE, INT_SOFTWARE_MACHINE, INT_TIMER_MACHINE,
-            INT_EXT_SUPER, INT_SOFTWARE_SUPER, INT_TIMER_SUPER,
-            INT_EXT_USER, INT_SOFTWARE_USER, INT_TIMER_USER
-        };
-        for (const int &id : interrupt_order)
-            if (checkInterrupt(id) && mask[id])
-                return std::make_shared<InterruptFault>(id);
-        return NoFault;
-    }
+    Fault getInterrupt() override;
 
     void updateIntrInfo() override {}
 
-    void
-    post(int int_num, int index) override
-    {
-        DPRINTF(Interrupt, "Interrupt %d:%d posted\n", int_num, index);
-        if (int_num != INT_NMI) {
-            ip[int_num] = true;
-        } else {
-            postNMI();
-        }
-    }
+    void post(int int_num, int index) override;
 
-    void
-    clear(int int_num, int index) override
-    {
-        DPRINTF(Interrupt, "Interrupt %d:%d cleared\n", int_num, index);
-        if (int_num != INT_NMI) {
-            ip[int_num] = false;
-        } else {
-            clearNMI();
-        }
-    }
+    void clear(int int_num, int index) override;
 
     void postNMI() { tc->setMiscReg(MISCREG_NMIP, 1); }
     void clearNMI() { tc->setMiscReg(MISCREG_NMIP, 0); }
 
-    void
-    clearAll() override
-    {
-        DPRINTF(Interrupt, "All interrupts cleared\n");
-        ip = 0;
-        clearNMI();
-    }
+    void clearAll() override;
 
     uint64_t readIP() const { return (uint64_t)ip.to_ulong(); }
     uint64_t readIE() const { return (uint64_t)ie.to_ulong(); }
     void setIP(const uint64_t& val) { ip = val; }
     void setIE(const uint64_t& val) { ie = val; }
 
-    void
-    serialize(CheckpointOut &cp) const override
-    {
-        unsigned long ip_ulong = ip.to_ulong();
-        unsigned long ie_ulong = ie.to_ulong();
-        SERIALIZE_SCALAR(ip_ulong);
-        SERIALIZE_SCALAR(ie_ulong);
-    }
+    void serialize(CheckpointOut &cp) const override;
 
-    void
-    unserialize(CheckpointIn &cp) override
-    {
-        unsigned long ip_ulong;
-        unsigned long ie_ulong;
-        UNSERIALIZE_SCALAR(ip_ulong);
-        ip = ip_ulong;
-        UNSERIALIZE_SCALAR(ie_ulong);
-        ie = ie_ulong;
-    }
+    void unserialize(CheckpointIn &cp) override;
+
+    Port &getPort(const std::string &if_name, PortID idx) override;
+
+    void raiseInterruptPin(uint32_t num);
+    void lowerInterruptPin(uint32_t num) {};
```

Comment:
I think the device cloud not clear the interrupt pending without handling the `lowerInterruptPin`, is right?

Reply:
I have no idea why you need to raise interrupt pins twice. IMHO, the gem5 will check interrupts after each instructions is commit. So we can just call `raise()` once in the interrupt source and the interrupt pending will be set. For cleaning the interrupt pending, it should be the process of ISR execution.

Just curious the interrupt pending will be clear before jumping the ISR?

Reply:
Please correct me if I understood your question wrong:

At the moment, local interrupt pending bits can only be cleared by writing to `mip`. To enable consecutive interrupts from devices with periodic interrupts, e.g., a sensor, `lowerInterruptPin()` has to be called before the next `raiseInterruptPin()`.

Reply:
Ok, I will check the issue again after long Chinese New Year holiday.